### DISCUSSION TITLE
Strict enums from typescript types

### DISCUSSION THREAD

**Discussion by wcastand** [colinhacks/zod#3549] [Q&A] [answered]
my use case is, i use `gql.tada` that generate types for my graphql api, but only generate types so i need to define the enum myself.

we use zod to define schema and validate our form, etc so i wanted to create an enum in zod that match and is validated by the enum from gql.tada.

code look like this: 

```
export type Frequency = ReturnType<typeof graphql.scalar<"Frequency">> // Frequency = "IMMEDIATE" | "ONCE" | "DAILY" | "WEEKLY" | "MONTHLY"

export const zFrequency: z.ZodType<Frequency> = z.enum(["IMMEDIATE", "ONCE", "DAILY", "WEEKLY"])
```

but in it's case i get no error from typescript while it should because zod enum is missing MONTHLY.

I tried different lib and superstruct has `Describe` which is the equivalent of `z.ZodType` that manage to give me the proper error in 

```
export const sFrequency: Describe<Frequency> = enums(["IMMEDIATE", "ONCE", "DAILY", "WEEKLY"])
/**
ERROR: 
[...]
Types of property 'schema' are incompatible.
    Property 'MONTHLY' is missing in type 
[...]
**/
```

was wondering if zod had an equivalent to `Describe` for a stricter match of enums?

EDIT:
just found a new issue exposing the same problem 
https://github.com/colinhacks/zod/issues/3534

**Comment by uncrft** [comment, marked as answer]
The `UnionToTuple` type from [type-fest](https://github.com/sindresorhus/type-fest) cat be used to ensure the zod enum type matches the GraphQL type from `gql.tada`.

```typescript 
import { type UnionToTuple } from "type-fest"
import { type graphql } from "gql.tada"
import { z } from "zod"

type Scalar<TScalar extends Readonly<string>> = ReturnType<typeof graphql.scalar<TScalar>>

const FrequencySchema = z.enum<string, UnionToTuple<Scalar<"Frequency">>>(
  ["IMMEDIATE", "ONCE", "DAILY", "WEEKLY", "MONTHLY"]
) // ✅

const FrequencySchema = z.enum<string, UnionToTuple<Scalar<"Frequency">>>(
  ["IMMEDIATE", "ONCE", "DAILY", "WEEKLY"]
) // ❌
```

Error for missing members should look something like this:
```
 Argument of type '["IMMEDIATE", "ONCE", "DAILY", "WEEKLY"]' is not assignable to parameter of type '["IMMEDIATE", "ONCE", "DAILY", "WEEKLY", "MONTHLY"]'.
```

Caveat: the array order needs to match the union order.

**Reply by wcastand** [reply]
i switched to effect/schema but same issue, i solved it using this: 

```typescript
import { assert, type Equals } from "tsafe"
assert<Equals<(typeof TransactionCategory)[number], TransactionCategory>>()
```
